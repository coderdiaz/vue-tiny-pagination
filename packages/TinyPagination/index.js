import TinyPagination from './src/main.vue';

TinyPagination.install = (Vue) => {
  Vue.component(TinyPagination.name, TinyPagination);
};

export default TinyPagination;
