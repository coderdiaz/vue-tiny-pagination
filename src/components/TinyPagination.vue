<template>
  <div class="vue-tiny-pagination">
    <ul class="tiny-pagination" :class="customClass">
      <li class="page-item" :class="classFirstPage">
        <a href="#" @click.prevent="lastPage">{{translation.prev}}</a>
      </li>
      <li class="page-item">
        <span>{{titlePage}}</span>
      </li>
      <li class="page-item" :class="classLastPage">
        <a href="#" @click.prevent="nextPage">{{translation.next}}</a>
      </li>
    </ul>
  </div>
</template>

<script>
import {Language} from '../utils'
export default {
  name: 'TinyPagination',
  props: {
    total: {
      type: Number,
      required: true
    },
    page: {
      type: Number,
      default: 1
    },
    lang: {
      type: String,
      default: 'en'
    },
    customClass: {
      type: String
    }
  },
  data () {
    return {
      currentPage: 1
    }
  },
  created () {
    this.currentPage = this.page
  },
  computed: {
    translation () {
      return Language.translations[this.lang]
    },
    totalPages () {
      return Math.ceil(this.total/10)
    },
    titlePage () {
      return `${this.translation.title} ${this.currentPage}`
    },
    classFirstPage () {
      return {
        disabled: (this.currentPage === 1),
        'c-not-allowed': true
      }
    },
    classLastPage () {
      return {
        disabled: (this.currentPage === this.totalPages),
        'c-not-allowed': true
      }
    }
  },
  methods: {
    nextPage () {
      if (this.currentPage !== this.totalPages) {
        this.currentPage += 1
      }
    },
    lastPage () {
      if (this.currentPage > 1) {
        this.currentPage -= 1
      }
    }
  },
  watch: {
    currentPage (value) {
      this.$emit('tiny:change-page', {
        page: value
      })
    }
  }
}
</script>

<style lang="scss">
// Pagination
.tiny-pagination {
  font-family: -apple-system, system-ui, BlinkMacSystemFont, "Segoe UI", Roboto, "Helvetica Neue", sans-serif;
  display: flex;
  list-style: none;
  margin: .2rem 0;
  padding: .2rem 0;

  .page-item {
    margin: .2rem .05rem;

    span {
      display: inline-block;
      padding: .2rem .2rem;
    }

    a {
      border-radius: .1rem;
      color: #667189;
      display: inline-block;
      padding: .2rem .4rem;
      text-decoration: none;
      &:focus,
      &:hover {
        color: #5755d9;
      }
    }

    &.disabled {
      a {
        cursor: default;
        opacity: .5;
        pointer-events: none;
      }
    }

    &.active {
      a {
        background: #5755d9;
        color: #fff;
      }
    }

    &.page-prev,
    &.page-next {
      flex: 1 0 50%;
    }

    &.page-next {
      text-align: right;
    }

    .page-item-title {
      margin: 0;
    }

    .page-item-subtitle {
      margin: 0;
      opacity: .5;
    }
  }
}
</style>