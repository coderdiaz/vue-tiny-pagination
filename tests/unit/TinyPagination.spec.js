import { shallowMount } from '@vue/test-utils';
import TinyPagination from '../../packages/TinyPagination/src/main.vue';

const vm = shallowMount(TinyPagination, {
  propsData: {
    total: 100,
  },
});

// Helper function to create a component
const createComponent = propsData => shallowMount(TinyPagination, { propsData });

describe('TinyPagination.vue', () => {
  let cmp;
  it('has a created hook', () => {
    expect(typeof TinyPagination.created).toBe('function');
  });

  it('should match the snapshot', () => {
    expect(vm.$el).toMatchSnapshot();
  });

  describe('Properties', () => {
    it('when the component is created without page prop, Page 1 is the page by default', () => {
      cmp = createComponent({ total: 300 });
      expect(cmp.vm.page).toBe(1);
    });

    it('when the page property is set, the currentPage is equals', () => {
      cmp = createComponent({ total: 300, page: 2 });
      expect(cmp.vm.currentPage).toBe(2);
    });

    it('when the component is created, English is the language by default', () => {
      cmp = createComponent({ total: 300 });
      expect(cmp.vm.lang).toBe('en');
      expect(cmp.vm.translation.title).toBe('Page');
    });

    it('when the lang prop is set to spanish, the component is translated', () => {
      cmp = createComponent({ total: 300, lang: 'es' });
      expect(cmp.vm.lang).toBe('es');
      expect(cmp.vm.translation.title).toBe('PÃ¡gina');
    });

    it('when the lang prop is set to not available language, English is the language by default', () => {
      cmp = createComponent({ total: 300, lang: 'fr' });
      expect(cmp.vm.translation.title).toBe('Page');
    });

    it('when the show limit is not set, true is by default', () => {
      cmp = createComponent({ total: 100 });
      expect(cmp.vm.showLimit).toBe(true);
    });

    it('when the showLimit is set, prop is same', () => {
      const showLimit = true;
      cmp = createComponent({ total: 100, showLimit });
      expect(cmp.vm.showLimit).toBe(showLimit);
    });
  });

  describe('Watchers', () => {
    it('currentPage watcher is called with the new value', () => {
      cmp = createComponent({ total: 100 });
      cmp.setData({ currentPage: 3 });
      expect(cmp.emitted()['tiny:change-page']).toBeTruthy();
    });

    it('currentLimit watcher is called with the new value', () => {
      cmp = createComponent({ total: 200 });
      cmp.setData({ currentLimit: 20 });
      expect(cmp.emitted()['tiny:change-limit']).toBeTruthy();
    });

    it('when the currentPage watcher is called, the tiny:change-page event is emitted', () => {
      const stub = jest.fn();
      cmp = createComponent({ total: 100 });
      cmp.vm.$on('tiny:change-page', stub);

      cmp.setData({ currentPage: 3 });
      expect(stub).toBeCalledWith({ page: 3 });
    });

    it('when the currentLimit watcher is called, the tiny:change-limit event is emitted', () => {
      const stub = jest.fn();
      cmp = createComponent({ total: 100 });
      cmp.vm.$on('tiny:change-limit', stub);

      cmp.setData({ currentLimit: 20 });
      expect(stub).toBeCalledWith({ limit: 20 });
    });
  });

  describe('Events', () => {
    beforeEach(() => {
      cmp = createComponent({ total: 20 });
    });

    it('calls nextPage when click on next button', () => {
      cmp.vm.nextPage = jest.fn();

      cmp.find('.btn-next-page').trigger('click');
      expect(cmp.vm.nextPage).toBeCalled();
    });

    it('call lastPage when click on prev button', () => {
      cmp.vm.lastPage = jest.fn();

      cmp.find('.btn-prev-page').trigger('click');
      expect(cmp.vm.lastPage).toBeCalled();
    });
  });
});
