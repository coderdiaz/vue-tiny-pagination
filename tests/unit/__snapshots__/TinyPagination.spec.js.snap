// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TinyPagination.vue should match the snapshot 1`] = `undefined`;
